polycommit/src/sonic_pc/gadgets/sonic_kzg10.rs

```diff
@@ -0,0 +1,705 @@
+        SonicKZG10,
+    },
+    BTreeMap,
+    BTreeSet,
+    EvaluationsVar,
+    LabeledPointVar,
+    LinearCombinationCoeffVar,
+    LinearCombinationVar,
+    PCCheckRandomDataVar,
+    PCCheckVar,
+    QuerySetVar,
+    String,
+    Vec,
+};
+
+/// Gadget for each entry for the linear combination
+#[derive(Derivative)]
+#[derivative(Clone(bound = "TargetCurve: PairingEngine,
+        BaseCurve: PairingEngine,
+        PG: PairingGadget<TargetCurve, <BaseCurve as PairingEngine>::Fr>"))]
+pub struct LCInfoEntry<TargetCurve, BaseCurve, PG>
+where
+    TargetCurve: PairingEngine,
+    BaseCurve: PairingEngine,
+    PG: PairingGadget<TargetCurve, <BaseCurve as PairingEngine>::Fr>,
+{
+    pub coeff: Option<NonNativeFieldVar<<TargetCurve as PairingEngine>::Fr, <BaseCurve as PairingEngine>::Fr>>,
+    pub degree_bound: Option<FpGadget<<BaseCurve as PairingEngine>::Fr>>,
+    pub prepared_comm: PreparedCommitmentVar<TargetCurve, BaseCurve, PG>,
+    pub negate: bool,
+}
+
+impl<TargetCurve, BaseCurve, PG> LCInfoEntry<TargetCurve, BaseCurve, PG>
+where
+    TargetCurve: PairingEngine,
+    BaseCurve: PairingEngine,
+    PG: PairingGadget<TargetCurve, <BaseCurve as PairingEngine>::Fr>,
+{
+    pub fn new(
+        coeff: &Option<NonNativeFieldVar<<TargetCurve as PairingEngine>::Fr, <BaseCurve as PairingEngine>::Fr>>,
+        degree_bound: &Option<FpGadget<<BaseCurve as PairingEngine>::Fr>>,
+        prepared_comm: &PreparedCommitmentVar<TargetCurve, BaseCurve, PG>,
+        negate: bool,
+    ) -> Self {
+        Self {
+            coeff: coeff.clone(),
+            degree_bound: degree_bound.clone(),
+            prepared_comm: prepared_comm.clone(),
+            negate,
+        }
+    }
+}
+
+/// Gadget for the Sonic-KZG10 polynomial commitment verifier.
+pub struct SonicKZG10Gadget<TargetCurve, BaseCurve, PG>
+where
+    TargetCurve: PairingEngine,
+    BaseCurve: PairingEngine,
+    PG: PairingGadget<TargetCurve, <BaseCurve as PairingEngine>::Fr>,
+{
+    _target_curve: PhantomData<TargetCurve>,
+    _base_curve: PhantomData<BaseCurve>,
+    _pairing_gadget: PhantomData<PG>,
+}
+
+impl<TargetCurve, BaseCurve, PG> Clone for SonicKZG10Gadget<TargetCurve, BaseCurve, PG>
+where
+    TargetCurve: PairingEngine,
+    BaseCurve: PairingEngine,
+    PG: PairingGadget<TargetCurve, <BaseCurve as PairingEngine>::Fr>,
+{
+    fn clone(&self) -> Self {
+        Self {
+            _target_curve: PhantomData,
+            _base_curve: PhantomData,
+            _pairing_gadget: PhantomData,
+        }
+    }
+}
+
+impl<TargetCurve, BaseCurve, PG> SonicKZG10Gadget<TargetCurve, BaseCurve, PG>
+where
+    TargetCurve: PairingEngine<Fq = <BaseCurve as PairingEngine>::Fr>,
+    BaseCurve: PairingEngine,
+    PG: PairingGadget<TargetCurve, <BaseCurve as PairingEngine>::Fr>,
+{
+    #[allow(clippy::type_complexity, clippy::too_many_arguments)]
+    fn prepared_batch_check_evaluations<CS: ConstraintSystem<<BaseCurve as PairingEngine>::Fr>>(
```

Comment:
this method is very long and complex - it would be great if it could be split into smaller chunks

Reply:
I may do it in a future PR. The upstream arkworks-rs/poly-commit is doing some similar refactoring aiming to simplify the arguments passing to the function. And we may follow up with those changes later.